src/utils/dockerContextManager.ts

```diff
@@ -0,0 +1,190 @@
+    // If three files are present, it is the CA certificate for Docker endpoint verification, client certificate and client key, in that order
+    // See https://docs.docker.com/engine/security/https/ for more details
+    // This is all quite complicated and VS Code Docker extension does not support these settings
+    [name: string]: string[]
+}
+
+export interface IDockerContextMetadata {
+    [name: string]: unknown
+}
+
+export interface IDockerContextStorage {
+    MetadataPath?: string,
+    TLSPath?: string
+}
+
+export interface IDockerContext {
+    Name: string,
+    Metadata: IDockerContextMetadata,
+    Endpoints: IDockerEndpoints,
+    TLSMaterial: IDockerTlsMaterial,
+    Storage: IDockerContextStorage,
+    FullSpec: string,       // Useful for checking if the context didn't change
+    HostProtocol?: string   // http, https, ssh, unix, npipe, ...
+}
+
+export interface IDockerContextCheckResult {
+    Context: IDockerContext,
+    Changed: boolean
+}
+
+export class DockerContextManager {
+    private readonly contextRefreshIntervalMs: number;
+    private lastContextCheckTimestamp: number;
+    private cachedContext: IDockerContext;
+    private lastDockerConfigDigest: string;
+    private contextChanged: boolean;
+
+    public constructor() {
+        this.lastContextCheckTimestamp = 0;
+        const configOptions: WorkspaceConfiguration = workspace.getConfiguration('docker');
+        this.contextRefreshIntervalMs = configOptions.get<number>('contextRefreshInterval', DefaultRefreshIntervalSec) * 1000;
+    }
+
+    public async getCurrentContext(): Promise<IDockerContextCheckResult> {
+        if (!this.cachedContext || (Date.now() - this.lastContextCheckTimestamp > this.contextRefreshIntervalMs)) {
+            try {
+                if (!this.cachedContext) {
+                    // First-time check
+                    this.lastDockerConfigDigest = await this.getDockerConfigDigest();
+                    await this.refreshCachedDockerContext();
+                } else {
+                    const dockerConfigDigest: string = await this.getDockerConfigDigest();
+
+                    if (!dockerConfigDigest || dockerConfigDigest !== this.lastDockerConfigDigest) {
+                        this.lastDockerConfigDigest = dockerConfigDigest;
+                        // Config file will change when Docker context is changed, but opposite is not necessarily true
+                        // (i.e. config file might change for other reasons).
+                        await this.refreshCachedDockerContext();
+                    }
+                }
+            } finally {
+                // Start counting time _after_ we are done with all the I/O
+                this.lastContextCheckTimestamp = Date.now();
+            }
+        }
+
+        return {
+            Context: this.cachedContext,
+            Changed: this.contextChanged
+        };
+    }
+
+    public contextChangeHandled(): void {
```

Comment:
This pattern feels a little strange to me...we're using a private variable to avoid passing a value around, and need to expose this method to callers in order to get that variable reset. If the callers forget I'm not sure what problems would result. Why not just return this from `refreshCachedDockerContext`?

Reply:
I think you are right, it is a bit over-engineered. Let me fix this ☺️